Sparse CSR: Add torch.empty

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66401
* #62180
* #65606
* #64391
* #61536
* #61858
* #61857
* #63948
* #63511
* #63510
* __->__ #63509
* #66485

The primary use of `torch.empty` is to reserve memory for tensor and set the type, device, size information. The same is done here for SparseCSR.
`crow_indices` is initialized as an empty tensor of size `num_rows + 1`. `col_indices` and `values` are initialized as empty tensors of size 0.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

Differential Revision: [D31770359](https://our.internmc.facebook.com/intern/diff/D31770359)

IvanYashchuk committed:
Add torch.empty for sparse csr layout

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add torch.empty for sparse csr layout"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -217,6 +217,38 @@ Tensor sparse_csr_tensor(
+  int64_t nnz = 0;
+
+  TensorOptions options = TensorOptions().dtype(ScalarType::Long).layout(Layout::Strided).device(device).pinned_memory(pin_memory);
+  auto crow_indices = at::empty({rows + 1}, options);
```
If you call empty for a strided Tensor you still get valid stride and size, but here empty would cause crow_indices and col_indices to be caused with values that could not be within bounds. Should we instead use something like arange to generate very basic, but valid indices for this?

IvanYashchuk replied:
Empty for strided Tensor has garbage as values. `crow_indices` must have the size `rows+1`, size of `crow_indices` changes only under resize operation, otherwise, it's fixed and its actual content is tied to values and column indices. I don't think it's a problem that `empty` returns crow_indices with garbage, it's used for allocating the memory and then filling it.

cpuhrsch replied:
Yes, the content of the memory for strided Tensor is garbage, but the size and stride fields are still valid. To some extend the indices here are extensions of these, since they're part of the storage as well. Even an empty strided Tensor won't run into segfaults if indexed for any sequence of valid indices. In this case however we might end up with elements in the row and col indices that cause segfaults when used to index into the values Tensor. So there exists an index for a sparse CSR tensor that causes a segfault, which does not exist for strided Tensors.

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"


The primary use of `torch.empty` is to reserve memory for tensor and set the type, device, size information. The same is done here for SparseCSR.
`crow_indices` is initialized as an empty tensor of size `num_rows + 1`. `col_indices` and `values` are initialized as empty tensors of size 0.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"


The primary use of `torch.empty` is to reserve memory for tensor and set the type, device, size information. The same is done here for SparseCSR.
`crow_indices` is initialized as an empty tensor of size `num_rows + 1`. `col_indices` and `values` are initialized as empty tensors of size 0.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add torch.empty"


The primary use of `torch.empty` is to reserve memory for tensor and set the type, device, size information. The same is done here for SparseCSR.
`crow_indices` is initialized as an empty tensor of size `num_rows + 1`. `col_indices` and `values` are initialized as empty tensors of size 0.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]
